<a href="https://colab.research.google.com/github/leiyi-ihb/ColabPractics/blob/main/B_practices/keras_example04_%E4%BA%A4%E5%8F%89%E7%86%B5%E5%92%8CDropout.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
# 这一节主要针对上一节的MNIST数据集分类进行优化调整
# 就只将模型中的loss="mse"改为了loss="categorical_crossentropy"
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD 

In [5]:
#输入数据
(x_train, y_train), (x_test, y_test)=mnist.load_data()
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

(60000, 28, 28) (60000,)
(10000, 28, 28) (10000,)


In [6]:
# 将(60000, 28, 28)转为(60000, 784)，这里的784其实就是28*28【我的理解应该是降维处理】
x_train = x_train.reshape(x_train.shape[0], -1) #行数就是取原来的60000，列数取-1表示不确定，由程序自己判断。 
# x_train = x_train.reshape(x_train.shape[0], 784) #跟上面一样的结果
x_test = x_test.reshape(x_test.shape[0], -1)
print(x_train.shape, len(x_train[0])) #可以发现，数据就变成一个60000行，784列的数据了
# print(x_train[0])

#需要对数据做个归一化
print(x_train.max()) 
x_train = x_train/255 #知道这里为什么除的是255吗？最大值为255
# print(x_train[0])
print(x_train.max())

x_test = x_test/255

# 转换label为one hot形式
from keras.utils import np_utils
print(y_train[0])
y_train = np_utils.to_categorical(y_train, num_classes=10) #这里为啥分为10个类没有讲清楚
y_test = np_utils.to_categorical(y_test, num_classes=10) #这里为啥分为10个类没有讲清楚
print(y_train[0])

(60000, 784) 784
255
1.0
5
[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]


In [7]:
# 创建模型
model = Sequential([
    Dense(units=10, input_dim=784, bias_initializer="one", activation="softmax")
])

# 定义优化器，编译模型。一定要注意optimizer别写成Optimizer或optimizers噢噢噢噢
sgd = SGD(lr=0.2)
model.compile(optimizer=sgd, loss="categorical_crossentropy", metrics= ['accuracy']) 

# 开始训练模型
model.fit(x_train, y_train, epochs=10, batch_size=32)

# 评估模型
loss, accuracy = model.evaluate(x_test, y_test)
print("loss:", loss)
print('accuracy:', accuracy)

/usr/local/lib/python3.7/dist-packages/keras/optimizers/optimizer_v2/gradient_descent.py:108: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3774 - accuracy: 0.8940
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3025 - accuracy: 0.9149
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2892 - accuracy: 0.9185
Epoch 4/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.2827 - accuracy: 0.9204
Epoch 5/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2780 - accuracy: 0.9223
Epoch 6/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2749 - accuracy: 0.9231
Epoch 7/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2712 - accuracy: 0.9247
Epoch 8/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2690 - accuracy: 0.9252
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2669 - accuracy: 0.9253
Epoch 10/10
313/313 [==============================] - 1s 3ms/step - loss: 0.2755 

**结果：**可以发现迭代第一次，准确度为0.8940，而上一节中，想要达到接近这个水平，约需要迭代3次，因此这里的效果好一些，但没有讲清楚怎么一下就知道使用这个的原因

# 接下来看看Dropout的使用

In [15]:
from keras.backend import dropout
from keras.layers.regularization.dropout import Dropout
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.optimizers import SGD

In [16]:
#输入数据
(x_train, y_train), (x_test, y_test)=mnist.load_data()
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

(60000, 28, 28) (60000,)
(10000, 28, 28) (10000,)


In [17]:
# 将(60000, 28, 28)转为(60000, 784)，这里的784其实就是28*28【我的理解应该是降维处理】
x_train = x_train.reshape(x_train.shape[0], -1) #行数就是取原来的60000，列数取-1表示不确定，由程序自己判断。 
# x_train = x_train.reshape(x_train.shape[0], 784) #跟上面一样的结果
x_test = x_test.reshape(x_test.shape[0], -1)
print(x_train.shape, len(x_train[0])) #可以发现，数据就变成一个60000行，784列的数据了
# print(x_train[0])

#需要对数据做个归一化
print(x_train.max()) 
x_train = x_train/255 #知道这里为什么除的是255吗？最大值为255
# print(x_train[0])
print(x_train.max())

x_test = x_test/255

# 转换label为one hot形式
from keras.utils import np_utils
print(y_train[0])
y_train = np_utils.to_categorical(y_train, num_classes=10) #这里为啥分为10个类没有讲清楚
y_test = np_utils.to_categorical(y_test, num_classes=10) #这里为啥分为10个类没有讲清楚
print(y_train[0])

(60000, 784) 784
255
1.0
5
[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]


In [20]:
# 创建模型
model = Sequential([
    Dense(units=200, input_dim=784, bias_initializer="one", activation="tanh"),
    Dense(units=100, bias_initializer="one", activation="tanh"),
    Dense(units=10, bias_initializer="one", activation="softmax")
])

# 定义优化器，编译模型。一定要注意optimizer别写成Optimizer或optimizers噢噢噢噢
sgd = SGD(lr=0.2)
model.compile(optimizer=sgd, loss="categorical_crossentropy", metrics= ['accuracy']) 

# 开始训练模型
model.fit(x_train, y_train, epochs=10, batch_size=32)

# 评估模型
loss, accuracy = model.evaluate(x_test, y_test)
print("\ntest loss:", loss)
print('accuracy:', accuracy)

loss, accuracy = model.evaluate(x_train, y_train) #将我们的训练集也评估一下
print("\ntrain loss:", loss)
print('accuracy:', accuracy)

Epoch 1/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2812 - accuracy: 0.9148
Epoch 2/10
1875/1875 [==============================] - 12s 7ms/step - loss: 0.1166 - accuracy: 0.9635
Epoch 3/10
1875/1875 [==============================] - 12s 7ms/step - loss: 0.0783 - accuracy: 0.9755
Epoch 4/10
1875/1875 [==============================] - 12s 7ms/step - loss: 0.0584 - accuracy: 0.9817
Epoch 5/10
1875/1875 [==============================] - 11s 6ms/step - loss: 0.0442 - accuracy: 0.9863
Epoch 6/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0323 - accuracy: 0.9898
Epoch 7/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.0243 - accuracy: 0.9927
Epoch 8/10
1875/1875 [==============================] - 12s 6ms/step - loss: 0.0177 - accuracy: 0.9949
Epoch 9/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0123 - accuracy: 0.9970
Epoch 10/10
313/313 [==============================] - 1s 4ms/step - loss:

**结果：**可以发现增加网络复杂度之后，准确率有很大的提升。测试集的准确度为0.9790，训练集的准确率为0.9993，很明显测试集同训练集之间存在一定的差距，因此有点过拟合了【其实就是判断两者差异】。其实我们这个例子差距不是很大，但以后很可能就会遇到差距大的，需要注意这一点。

In [21]:
# 创建模型
model = Sequential([
    Dense(units=200, input_dim=784, bias_initializer="one", activation="tanh"),
    Dropout(0.4), #其实就是让上面的神经元的40%不工作
    Dense(units=100, bias_initializer="one", activation="tanh"),
    Dropout(0.4),
    Dense(units=10, bias_initializer="one", activation="softmax")
])

# 定义优化器，编译模型。一定要注意optimizer别写成Optimizer或optimizers噢噢噢噢
sgd = SGD(lr=0.2)
model.compile(optimizer=sgd, loss="categorical_crossentropy", metrics= ['accuracy']) 

# 开始训练模型
model.fit(x_train, y_train, epochs=10, batch_size=32)

# 评估模型
loss, accuracy = model.evaluate(x_test, y_test)
print("\ntest loss:", loss)
print('accuracy:', accuracy)

loss, accuracy = model.evaluate(x_train, y_train) #将我们的训练集也评估一下
print("\ntrain loss:", loss)
print('accuracy:', accuracy)

Epoch 1/10
1875/1875 [==============================] - 9s 4ms/step - loss: 0.4485 - accuracy: 0.8640
Epoch 2/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2877 - accuracy: 0.9157
Epoch 3/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.2423 - accuracy: 0.9288
Epoch 4/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2161 - accuracy: 0.9364
Epoch 5/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.1981 - accuracy: 0.9409
Epoch 6/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.1880 - accuracy: 0.9441
Epoch 7/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1760 - accuracy: 0.9478
Epoch 8/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.1701 - accuracy: 0.9497
Epoch 9/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.1594 - accuracy: 0.9513
Epoch 10/10
313/313 [==============================] - 1s 2ms/step - loss: 0.09

**结果：**在这个例子中可以发现，测试集最后的准确度还不如上次高了【说明次例子并不适合dropout，只有凑合这讲】，但Droput使得测试和训练准确度间的差异确实缩小了一些，但是还是还是存在些许过拟合（所以这个并非是讲解Dropout的最好例子）  
还有一个特点可以发现，训练过程中出现的准确度(最高为0.9544)都是比最终出来的准确度(0.9706)小的，这是因为训练过程使用Dropout,因此并没有使用全神经元做判断，而最后的准确度计算是包含全神经元的。